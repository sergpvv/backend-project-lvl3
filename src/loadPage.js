import { promises as fs, createWriteStream } from 'fs';
import axios from 'axios';
import { parse, resolve } from 'url';
import cheerio from 'cheerio';
import debug from 'debug';
import Listr from 'listr';
import getLocalNames, { isLocal, buildResourceFilename } from './utils';

const log = debug('page-loader');

const resourcesMap = { link: 'href', script: 'src', img: 'src' };

const download = ({ resource, filename }) => {
  log(`download ${resource}`);
  // console.log(`resource: ${resource}\nfilename: ${filename}`);
  return axios
    .get(resource, { responseType: 'stream' })
    .then(({ data }) => data.pipe(createWriteStream(filename)))
    .then(() => log(`${resource} successfully saved as ${filename}`));
};

const processResources = (html, pageUrl, outputDirectory) => {
  const { hostname } = parse(pageUrl);
  const dom = cheerio.load(html, { decodeEntities: false });
  /* const localResources = [];
  Object.entries(resourcesMap)
    .forEach(([tag, src]) => {
      dom(`${tag}[${src}]`).each((index, element) => {
        const resource = dom(element).attr(src);
        log(`processing ${resource}`);
        if (isLocal(resource, hostname)) {
          const filename = buildResourceFilename(outputDirectory, resource);
          dom(element).attr(src, filename);
          localResources.push({ resource: resolve(pageUrl, resource), filename });
        }
      });
    });
*/
  const getUrl = ({ name, attribs }) => attribs[resourcesMap[name]];
  const localResources = Object.entries(resourcesMap)
    .map(([tag, src]) => dom(`${tag}[${src}]`).toArray())
    .flat()
    .filter((element) => isLocal(getUrl(element), hostname))
    .map((element) => {
      const url = getUrl(element);
      log(`processing ${url}`);
      const resource = resolve(pageUrl, url);
      const filename = buildResourceFilename(outputDirectory, url);
      dom(element).attr(resourcesMap[element.name], filename);
      return { resource, filename };
    });
  log('resources processing is complete');
  const pageWithLocalRes = dom.html();
  return { pageWithLocalRes, localResources };
};

const saveDownloadedPage = (filename, data) => {
  log(`save downloaded page to ${filename}`);
  return fs.writeFile(filename, data, 'utf-8');
};

const downloadResources = (resources, pageUrl, resourcesDirectory) => {
  log(`make directory ${resourcesDirectory} `);
  return fs.mkdir(resourcesDirectory)
    .then(() => {
      const listrTasks = new Listr(
        resources.map(({ resource, filename }) => ({
          title: resource,
          task: () => download({ resource, filename }),
        })),
        { concurrent: true, exitOnError: false },
      );
      log('download local resources');
      return listrTasks.run();
    });
};

export default (pageUrl, outputDirectory) => {
  const { filename, resourcesDirectory } = getLocalNames(pageUrl, outputDirectory);
  return axios
    .get(pageUrl)
    .then(({ data }) => {
      log(`${pageUrl} download complete, processing resources`);
      return processResources(data, pageUrl, resourcesDirectory);
    })
    .then(({ pageWithLocalRes, localResources }) => Promise.all([
      saveDownloadedPage(filename, pageWithLocalRes),
      downloadResources(localResources, pageUrl, resourcesDirectory),
    ]))
    .then(() => {
      log(`${pageUrl} successfully saved as ${filename}`);
      return filename;
    });
};
