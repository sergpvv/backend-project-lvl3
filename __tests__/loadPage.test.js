import { promises as fs } from 'fs';
import path from 'path';
import url from 'url';
import os from 'os';
import nock from 'nock';
import loadPage, { normalize } from '../src';

const buildFilename = (name) => path.join(__dirname, '__fixtures__', name);
const makeTmpDir = () => fs.mkdtemp(path.join(os.tmpdir(), path.sep));
const join = (...strParts) => strParts.join('');

const testSite = 'https://tests.ite';
const pagePath = '/test/page';
const testPage = join(testSite, pagePath);

const getPageName = (pageUrl) => {
  const { hostname, pathname } = url.parse(pageUrl);
  return normalize(path.join(hostname, pathname));
};
const pageName = getPageName(testPage);
const pageFilename = join(pageName, '.html');

const resourcesPath = '/assets';
const resourcesDirName = join(pageName, '_files');
const getResourceLocalName = (resourcePath, resourceName, resourceType) => {
  const resource = path.join(resourcePath, join(resourceName, resourceType));
  return normalize(resource.slice(1));
};

const resources = [
  ['helloworld', 'text/html', ''],
  ['style', 'text/css', '.css'],
  ['script', 'application/javascript', '.js'],
  ['picture', 'image/jpg', '.jpg'],
];

const nonexist = '/nonexistentresource';
const unknown = '/unknownhostname';

nock.disableNetConnect();

beforeEach(() => {
  const scope = nock(testSite);
  resources.map(([name, contentType, type]) => ({
    resource: type.length === 0 ? pagePath : path.join(resourcesPath, join(name, type)),
    filename: buildFilename(name),
    contentType,
  }))
    .forEach(({ resource, filename, contentType }) => {
      scope.get(resource).replyWithFile(200, filename, { 'Content-Type': contentType });
    });
  // console.log(scope.pendingMocks());
  scope
    .get(nonexist).reply(404)
    .get(unknown).replyWithError({
      message: 'getaddrinfo ENOTFOUND',
      code: 'ENOTFOUND',
    });
});

it('normalize string', () => {
  expect(normalize('fo.o/bar/ba.z')).toBe('fo-o-bar-ba-z');
});

it('correct data', async () => {
  const outputDirectory = await makeTmpDir();
  const filename = await loadPage(testPage, outputDirectory);
  const expectedFilename = path.join(outputDirectory, pageFilename);
  expect(await fs.readFile(filename, 'utf-8')).not.toBeNull();
  expect(filename).toBe(expectedFilename);

  [['style', '.css', 'utf-8'], ['script', '.js', 'utf-8']]
    .forEach(async ([name, type, encoding]) => {
      const resourceFilename = path.join(
        outputDirectory,
        resourcesDirName,
        getResourceLocalName(resourcesPath, name, type),
      );
      expect(await fs.readFile(resourceFilename, encoding))
        .toBe(await fs.readFile(buildFilename(name), encoding));
    });

  const { size: expected } = await fs.stat(buildFilename('picture'));
  const resourceFilename = path.join(
    outputDirectory,
    resourcesDirName,
    getResourceLocalName(resourcesPath, 'picture', '.jpg'),
  );
  const { size: actual } = await fs.stat(resourceFilename);
  expect(actual).toBe(expected);
});

it.each([['unknown hostname', join(testSite, unknown), null],
  ['nonexistent resource', join(testSite, nonexist), null],
  ['nonexistent output directory', testPage, nonexist],
  ['permission denied', testPage, '/']])(
  '%s',
  async (test, page, dir) => {
    const outputDir = dir === null ? await makeTmpDir() : dir;
    await expect(loadPage(page, outputDir))
      .rejects.toThrowErrorMatchingSnapshot();
  },
);
