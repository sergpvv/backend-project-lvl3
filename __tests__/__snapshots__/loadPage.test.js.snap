// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nonexistent output directory 1`] = `"ENOENT: no such file or directory, open '/nonexistentresource/tests-ite-test-page.html'"`;

exports[`nonexistent resource 1`] = `"Request failed with status code 404"`;

exports[`permission denied 1`] = `"EACCES: permission denied, open '/tests-ite-test-page.html'"`;

exports[`unknown hostname 1`] = `"getaddrinfo ENOTFOUND"`;
